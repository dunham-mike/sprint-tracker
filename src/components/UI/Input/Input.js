import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import Datetime from 'react-datetime';
import * as moment from 'moment';

import './Input.css';

const styles = theme => ({
    Input: {
        width: '100%',
        padding: '10px',
        boxSizing: 'border-box',
    },
    Label: {
        fontWeight: 'bold',
        display: 'block',
        marginBottom: '8px',
    },
    InputElement: {
        outline: 'none',
        border: '1px solid #ccc',
        backgroundColor: 'white',
        font: 'inherit',
        padding: '6px 10px',
        display: 'block',
        width: '100%',
        boxSizing: 'border-box',
    },
    
    // InputElement:focus: {
    //     outline: 'none'
    //     backgroundColor: '#ccc',
    // },
    
    Invalid: {
        border: '1px solid red',
        backgroundColor: '#FDA49A'
    },
});

const input = (props) => {
    const { classes } = props;

    let inputElement = null;
    const inputClasses = [classes.InputElement];

    if (!props.valid && props.shouldValidate && props.touched) {
        inputClasses.push(classes.Invalid);
    }

    switch (props.elementType) {
        case('readonly'):
            inputElement = (<div>{props.value}</div>);
            break;
        case('input'):
            inputElement = <input 
                className={inputClasses.join(' ')} 
                value={props.value}
                onChange={props.changed}/>;
            break;
        case ('textarea'):
            inputElement = <textarea 
                className={inputClasses.join(' ')} 
                rows='3'
                value={props.value}
                onChange={props.changed}/>;
            break;
        case ('select'):
            inputElement = (
                <select
                    className={inputClasses.join(' ') + ' Select'}
                    value={props.value}
                    style={{width: 'auto'}}
                    onChange={props.changed}>
                    {props.elementConfig.options.map(option => (
                        <option key={option.value} value={option.value}>
                            {option.displayValue}
                        </option>
                    ))}
                </select>);
            break;
        case ('date'):
            inputElement = (
                <Datetime 
                    utc={true}
                    value={(props.value !== '') ? moment.utc(props.value).format("ddd. MMM. D, YYYY") : null}
                    viewDate={props.value}
                    onChange={props.dateChanged}
                    closeOnSelect={true}
                    timeFormat={false}
                    inputProps={{readOnly: true, className: "customInput"}} // Per: https://github.com/YouCanBookMe/react-datetime/issues/46
                />
            );
            break;
        default:
            inputElement = <input 
                className={inputClasses.join(' ')} 
                value={props.value}
                onChange={props.changed}/>;
    }

    // console.log('props.validation', props.validation);

    return (
        <div 
            className={classes.Input}
            >
            <label 
                className={classes.Label}
            >
                {props.elementConfig.displayName}
                {props.required ? '*' : null}
            </label>
            {inputElement}
        </div>
    );
}



export default withStyles(styles)(input);