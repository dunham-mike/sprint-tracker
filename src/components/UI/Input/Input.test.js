import React from 'react';
import renderer from 'react-test-renderer';
import * as moment from 'moment';

import Input from './Input';

describe('Render Input', () => {

    it('renders Input component correctly with no props', () => {  
        const InputComponent = renderer.create(<Input />).toJSON();
        expect(InputComponent).toMatchSnapshot();
    });

    it('renders Input component as "readonly" correctly', () => {  
        const props = {
            elementType: "readonly",
            value: "readonly value",
        };

        const InputComponent = renderer.create(<Input {...props} />).toJSON();
        expect(InputComponent).toMatchSnapshot();
    });

    it('renders Input component as "input" correctly', () => {  
        const props = {
            elementType: "input",
            value: "input value",
        };

        const InputComponent = renderer.create(<Input {...props} />).toJSON();
        expect(InputComponent).toMatchSnapshot();
    });

    it('renders Input component as "textarea" correctly', () => {  
        const props = {
            elementType: "textarea",
            value: "textarea value",
        };

        const InputComponent = renderer.create(<Input {...props} />).toJSON();
        expect(InputComponent).toMatchSnapshot();
    });

    it('renders Input component as "select" correctly', () => {  
        const props = {
            elementType: "select",
            value: "initial select value",
            elementConfig: {
                options: [
                    {value: 'initial select value', displayValue: 'Initial Select Value'},
                    {value: 'second select value', displayValue: 'Second Select Value'},
                ]
            }
        };

        const InputComponent = renderer.create(<Input {...props} />).toJSON();
        expect(InputComponent).toMatchSnapshot();
    });

    it('renders Input component as "date" correctly', () => {  
        const props = {
            elementType: "date",
            value: moment.utc(0),
        };

        const InputComponent = renderer.create(<Input {...props} />).toJSON();
        expect(InputComponent).toMatchSnapshot();
    });

});