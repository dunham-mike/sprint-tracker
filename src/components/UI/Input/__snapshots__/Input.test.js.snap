// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Input renders Input component as "date" correctly 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  >
    
  </label>
  <div
    className="rdt"
  >
    <input
      className="customInput"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      readOnly={true}
      type="text"
      value="Thu. Jan. 1, 1970"
    />
    <div
      className="rdtPicker"
    >
      <div
        className="rdtDays"
      >
        <table>
          <thead>
            <tr>
              <th
                className="rdtPrev"
                onClick={[Function]}
              >
                <span>
                  ‹
                </span>
              </th>
              <th
                className="rdtSwitch"
                colSpan={5}
                data-value={0}
                onClick={[Function]}
              >
                January 1970
              </th>
              <th
                className="rdtNext"
                onClick={[Function]}
              >
                <span>
                  ›
                </span>
              </th>
            </tr>
            <tr>
              <th
                className="dow"
              >
                Su
              </th>
              <th
                className="dow"
              >
                Mo
              </th>
              <th
                className="dow"
              >
                Tu
              </th>
              <th
                className="dow"
              >
                We
              </th>
              <th
                className="dow"
              >
                Th
              </th>
              <th
                className="dow"
              >
                Fr
              </th>
              <th
                className="dow"
              >
                Sa
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                className="rdtDay rdtOld"
                data-value={28}
                onClick={[Function]}
              >
                28
              </td>
              <td
                className="rdtDay rdtOld"
                data-value={29}
                onClick={[Function]}
              >
                29
              </td>
              <td
                className="rdtDay rdtOld"
                data-value={30}
                onClick={[Function]}
              >
                30
              </td>
              <td
                className="rdtDay rdtOld"
                data-value={31}
                onClick={[Function]}
              >
                31
              </td>
              <td
                className="rdtDay"
                data-value={1}
                onClick={[Function]}
              >
                1
              </td>
              <td
                className="rdtDay"
                data-value={2}
                onClick={[Function]}
              >
                2
              </td>
              <td
                className="rdtDay"
                data-value={3}
                onClick={[Function]}
              >
                3
              </td>
            </tr>
            <tr>
              <td
                className="rdtDay"
                data-value={4}
                onClick={[Function]}
              >
                4
              </td>
              <td
                className="rdtDay"
                data-value={5}
                onClick={[Function]}
              >
                5
              </td>
              <td
                className="rdtDay"
                data-value={6}
                onClick={[Function]}
              >
                6
              </td>
              <td
                className="rdtDay"
                data-value={7}
                onClick={[Function]}
              >
                7
              </td>
              <td
                className="rdtDay"
                data-value={8}
                onClick={[Function]}
              >
                8
              </td>
              <td
                className="rdtDay"
                data-value={9}
                onClick={[Function]}
              >
                9
              </td>
              <td
                className="rdtDay"
                data-value={10}
                onClick={[Function]}
              >
                10
              </td>
            </tr>
            <tr>
              <td
                className="rdtDay"
                data-value={11}
                onClick={[Function]}
              >
                11
              </td>
              <td
                className="rdtDay"
                data-value={12}
                onClick={[Function]}
              >
                12
              </td>
              <td
                className="rdtDay"
                data-value={13}
                onClick={[Function]}
              >
                13
              </td>
              <td
                className="rdtDay"
                data-value={14}
                onClick={[Function]}
              >
                14
              </td>
              <td
                className="rdtDay"
                data-value={15}
                onClick={[Function]}
              >
                15
              </td>
              <td
                className="rdtDay"
                data-value={16}
                onClick={[Function]}
              >
                16
              </td>
              <td
                className="rdtDay"
                data-value={17}
                onClick={[Function]}
              >
                17
              </td>
            </tr>
            <tr>
              <td
                className="rdtDay"
                data-value={18}
                onClick={[Function]}
              >
                18
              </td>
              <td
                className="rdtDay"
                data-value={19}
                onClick={[Function]}
              >
                19
              </td>
              <td
                className="rdtDay"
                data-value={20}
                onClick={[Function]}
              >
                20
              </td>
              <td
                className="rdtDay"
                data-value={21}
                onClick={[Function]}
              >
                21
              </td>
              <td
                className="rdtDay"
                data-value={22}
                onClick={[Function]}
              >
                22
              </td>
              <td
                className="rdtDay"
                data-value={23}
                onClick={[Function]}
              >
                23
              </td>
              <td
                className="rdtDay"
                data-value={24}
                onClick={[Function]}
              >
                24
              </td>
            </tr>
            <tr>
              <td
                className="rdtDay"
                data-value={25}
                onClick={[Function]}
              >
                25
              </td>
              <td
                className="rdtDay"
                data-value={26}
                onClick={[Function]}
              >
                26
              </td>
              <td
                className="rdtDay"
                data-value={27}
                onClick={[Function]}
              >
                27
              </td>
              <td
                className="rdtDay"
                data-value={28}
                onClick={[Function]}
              >
                28
              </td>
              <td
                className="rdtDay"
                data-value={29}
                onClick={[Function]}
              >
                29
              </td>
              <td
                className="rdtDay"
                data-value={30}
                onClick={[Function]}
              >
                30
              </td>
              <td
                className="rdtDay"
                data-value={31}
                onClick={[Function]}
              >
                31
              </td>
            </tr>
            <tr>
              <td
                className="rdtDay rdtNew"
                data-value={1}
                onClick={[Function]}
              >
                1
              </td>
              <td
                className="rdtDay rdtNew"
                data-value={2}
                onClick={[Function]}
              >
                2
              </td>
              <td
                className="rdtDay rdtNew"
                data-value={3}
                onClick={[Function]}
              >
                3
              </td>
              <td
                className="rdtDay rdtNew"
                data-value={4}
                onClick={[Function]}
              >
                4
              </td>
              <td
                className="rdtDay rdtNew"
                data-value={5}
                onClick={[Function]}
              >
                5
              </td>
              <td
                className="rdtDay rdtNew"
                data-value={6}
                onClick={[Function]}
              >
                6
              </td>
              <td
                className="rdtDay rdtNew"
                data-value={7}
                onClick={[Function]}
              >
                7
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render Input renders Input component as "input" correctly 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  >
    
  </label>
  <input
    className="input-InputElement-3"
    value="input value"
  />
</div>
`;

exports[`Render Input renders Input component as "readonly" correctly 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  >
    
  </label>
  <div
    data-testid="readonly"
  >
    readonly value
  </div>
</div>
`;

exports[`Render Input renders Input component as "select" correctly 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  />
  <select
    className="input-InputElement-3 Select"
    style={
      Object {
        "width": "auto",
      }
    }
    value="initial select value"
  >
    <option
      value="initial select value"
    >
      Initial Select Value
    </option>
    <option
      value="second select value"
    >
      Second Select Value
    </option>
  </select>
</div>
`;

exports[`Render Input renders Input component as "textarea" correctly 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  >
    
  </label>
  <textarea
    className="input-InputElement-3"
    rows="3"
    value="textarea value"
  />
</div>
`;

exports[`Render Input renders Input component correctly with no props 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  >
    
  </label>
  <input
    className="input-InputElement-3"
  />
</div>
`;

exports[`Render Input renders Input component when it is invalid correctly 1`] = `
<div
  className="input-Input-1"
>
  <label
    className="input-Label-2"
  >
    
  </label>
  <input
    className="input-InputElement-3 input-Invalid-4"
    value="input value"
  />
</div>
`;
